File: OrdoRubim.swift
/**
 * ════════════════════════════════════════════════════════════════════════════
 * ORDO RUBIM — SYMPHONIA CODICIS (SWIFT VERSION)
 * LIBRETTO: MATERIA OBSCURA
 * PLATFORM: iOS / MACOS / METAL
 * ════════════════════════════════════════════════════════════════════════════
 */

import Foundation

// [ACT I: VOCATIO ELEMENTORUM]

// ─── CANTUS PRIMUS: CHAOS ET ORDO ───────────────────────────────────────────

enum Chaos {
    // // Hic sunt dracones.
    struct RawInput {
        let data: String
        
        // // Nihil ex nihilo fit.
        var isVoid: Bool {
            return data.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
    }

    enum EntropyFilter {
        // // Purgamentum init, purgamentum exit.
        static func purge(_ input: RawInput) throws {
            // // Si silentium est...
            guard !input.isVoid else {
                // // ...clamor erroris audietur!
                throw NSError(domain: "ABYSSUS", code: 404, userInfo: [NSLocalizedDescriptionKey: "Vacuitas non accipitur."])
            }
        }
    }
}

// ─── CANTUS SECUNDUS: LEX AETERNA ───────────────────────────────────────────

enum Dictatura {
    enum Accuratio {
        // // Mensura sine errore.
        static func verifyPurity(_ value: Double) throws {
            // // Numerus insanus vetatur.
            if value.isNaN || value.isInfinite {
                throw NSError(domain: "HERESIS", code: 500, userInfo: [NSLocalizedDescriptionKey: "Numerus non existit."])
            }
        }

        // // Quod imperfectum est, corrigetur.
        static func snapToGrid(_ value: Double) -> Double {
            let limen = 1e-6
            return abs(value) < limen ? 0.0 : value
        }
    }

    actor ImmutableLaw {
        // // Unus rex, una lex.
        func acquireControl(action: () -> Void) {
            // // Ianua clauditur. 
            action()
            // // Ianua aperitur.
        }
    }
}

// ─── CANTUS TERTIUS: MACHINATIO ─────────────────────────────────────────────

enum MateriaObscura {
    enum VoidEngine {
        // // Et verbum caro factum est.
        static func invoke(_ will: String) {
            // // Machina obsequitur.
            print("[RUBIM] >>> Fiat: \(will)")
        }
    }
}

// ─── CANTUS QUARTUS: MAGISTER ───────────────────────────────────────────────

final class Magister {
    // // Ego sum Alpha et Omega.
    static let shared = Magister()
    
    private let law = Dictatura.ImmutableLaw()

    private init() {}

    // // In manu mea potestas est. 
    // Async function needed to enter the Actor's domain
    func dominariMateria(_ intentio: String) async throws {
        
        // // 1. INTROITUS (Вход)
        let input = Chaos.RawInput(data: intentio)

        // // Si indignus es, discede. 
        try Chaos.EntropyFilter.purge(input)

        // // 2. SILENTIUM 
        // // Tempus sistit. 
        await law.acquireControl {
            // // 3. ACTIO 
            // // Mundus mutatur. 
            MateriaObscura.VoidEngine.invoke(intentio)
        }
        
        // // 4. EXPLICIT 
    }
}

// ─── CANTUS QUINTUS: ADYTUM ─────────────────────────────────────────────────

protocol Initiatus {
    // // Nomen est omen. 
    static var nomenUmbrae: String { get }

    // // Opus perfectum requiritur. 
    init()
    func opusMagnum(_ guru: Magister) async throws
}

enum Adytum {
    static func openGates<T: Initiatus>(for adeptType: T.Type) {
        let nomen = T.nomenUmbrae
        
        // // Sine nomine, sine vita. 
        guard !nomen.isEmpty else {
            fatalError("DAMNATIO: Nomen vacuum.")
        }

        print("\n>>> GLORIA IN EXCELSIS CODEX.")
        print(">>> Adventus: \(nomen).\n")

        Task {
            do {
                let guru = Magister.shared
                let neophyte = T()
                
                // // Incipit Tragoedia. 
                try await neophyte.opusMagnum(guru)
                
                print("\n>>> CONSUMMATUM EST. (Свершилось.)")
                exit(0) // Successful exit

            } catch {
                // // Dies irae, dies illa. 
                print("\n!!! LACRIMOSA: \(error.localizedDescription)")
                exit(0xBADC0DE)
            }
        }
        
        // Keep main thread alive for async task (Command Line Tool specific)
        RunLoop.main.run()
    }
}

// ─── MACRO ACTIVATIONIS ─────────────────────────────────────────────────────

func InUmbraIgiturPugnabimus<T: Initiatus>(_ type: T.Type) {
    Adytum.openGates(for: type)
}

// ════════════════════════════════════════════════════════════════════════════
// EXEMPLUM 
// ════════════════════════════════════════════════════════════════════════════

/*
struct Architectus: Initiatus {
    static let nomenUmbrae = "Architect_iOS"

    func opusMagnum(_ guru: Magister) async throws {
        try await guru.dominariMateria("Fiat Lux")
    }
}

// InUmbraIgiturPugnabimus(Architectus.self)
*/

File: OrdoRubim.rs
// ════════════════════════════════════════════════════════════════════════════
//  ORDO RUBIM — SYMPHONIA CODICIS
//  LIBRETTO: MATERIA OBSCURA
//  EDITION: 2024 (RUST 1.80+)
// ════════════════════════════════════════════════════════════════════════════

// [ACT I: VOCATIO ELEMENTORUM]

use std::sync::{Mutex, MutexGuard, LazyLock}; // CUSTOS
use std::process;                             // EXITUS

// ─── CANTUS PRIMUS: CHAOS ET ORDO ────────────

mod chaos {
    // // Hic sunt dracones.
    pub struct RawInput<'a> {
        pub data: &'a str,
    }

    impl<'a> RawInput<'a> {
        // // Nihil ex nihilo fit.
        pub fn is_void(&self) -> bool {
            self.data.is_empty()
        }
    }

    pub struct EntropyFilter;

    impl EntropyFilter {
        // // Purgamentum init, purgamentum exit.
        pub fn purge(input: &RawInput) -> Result<(), &'static str> {
            // // Si silentium est...
            if input.is_void() {
                // // ...clamor erroris audietur!
                return Err("ABYSSUS: Vacuitas non accipitur.");
            }
            Ok(())
        }
    }
}

// ─── CANTUS SECUNDUS: LEX AETERNA ──────────────

mod dictatura {
    use std::sync::{Mutex, MutexGuard};

    pub struct Accuratio;

    impl Accuratio {
        // // Mensura sine errore.
        pub fn verify_purity(value: f64) -> Result<(), &'static str> {
            // // Numerus insanus vetatur.
            if value.is_nan() || value.is_infinite() {
                return Err("HERESIS: Numerus non existit.");
            }
            Ok(())
        }

        // // Quod imperfectum est, corrigetur.
        pub fn snap_to_grid(value: f64) -> f64 {
            const LIMEN: f64 = 1e-6; 
            if value.abs() < LIMEN {
                return 0.0;
            }
            value
        }
    }

    pub struct ImmutableLaw;

    // // Unus rex, una lex.
    static GATE_KEEPER: Mutex<()> = Mutex::new(());

    impl ImmutableLaw {
        pub fn acquire_control() -> MutexGuard<'static, ()> {
            // // Ianua clauditur.
            GATE_KEEPER.lock().unwrap_or_else(|p| p.into_inner())
        }
    }
}

// ─── CANTUS TERTIUS: MACHINATIO ────────────────────

mod materia_obscura {
    pub struct VoidEngine;

    impl VoidEngine {
        // // Et verbum caro factum est.
        pub fn invoke(will: &str) {
            // // Machina obsequitur.
            println!("[RUBIM] >>> Fiat: {}", will);
        }
    }
}

// ─── CANTUS QUARTUS: MAGISTER ────────────────────

use chaos::{RawInput, EntropyFilter};
use dictatura::ImmutableLaw;
use materia_obscura::VoidEngine;

// // Ego sum Alpha et Omega.
// Singleton через Unit Struct + Static State
pub struct Magister;

impl Magister {
    // // In manu mea potestas est. 
    pub fn dominari_materia(&self, intentio: &str) {
        
        // // 1. INTROITUS (Вход)
        let input = RawInput { data: intentio };

        // // Si indignus es, discede. 
        if let Err(e) = EntropyFilter::purge(&input) {
            panic!("ANATHEMA: {}", e);
        }

        // // 2. SILENTIUM 
        // // Tempus sistit. 
        let _guard = ImmutableLaw::acquire_control();

        // // 3. ACTIO 
        // // Mundus mutatur. 
        VoidEngine::invoke(intentio);

        // // 4. EXPLICIT 
        // // Pax vobiscum. (Mutex unlocks here automatically when _guard is dropped)
    }
}

// ─── CANTUS QUINTUS: ADYTUM ────────────────────────

// Trait вместо Concept/Interface
pub trait Initiatus {
    // // Nomen est omen. 
    // Associated Constant 
    const NOMEN_UMBRAE: &'static str;

    // // Opus perfectum requiritur. 
    fn opus_magnum(guru: &Magister);
}

pub struct Adytum;

impl Adytum {
    pub fn open_gates<T: Initiatus + Default>() {
        // // Nomina sunt odiosa? Non hic. 
        let nomen = T::NOMEN_UMBRAE;

        // // Sine nomine, sine vita. 
        if nomen.is_empty() {
             eprintln!("DAMNATIO: Nomen vacuum.");
             process::exit(0xDEAD);
        }

        println!("\n>>> GLORIA IN EXCELSIS CODEX.");
        println!(">>> Adventus: {}.\n", nomen);
        
        let result = std::panic::catch_unwind(|| {
            let guru = Magister;
            let neophyte = T::default();
            
            // // Incipit Tragoedia. 
            T::opus_magnum(&neophyte, &guru);
        });

        match result {
            Ok(_) => println!("\n>>> CONSUMMATUM EST. (Свершилось.)"),
            Err(_) => {
                // // Dies irae, dies illa. 
                eprintln!("\n!!! LACRIMOSA: Ritual Failed.");
                process::exit(0xBADC0DE);
            }
        }
    }
}

// ─── MACRO ACTIVATIONIS ────────────────────────────

#[macro_export]
macro_rules! in_umbra_igitur_pugnabimus {
    ($adept_type:ty) => {
        fn main() {
            Adytum::open_gates::<$adept_type>();
        }
    };
}

// ════════════════════════════════════════════════════════════════════════════
// EXEMPLUM 
// ════════════════════════════════════════════════════════════════════════════

/*
#[derive(Default)]
struct Architectus;

impl Initiatus for Architectus {
    const NOMEN_UMBRAE: &'static str = "Architect_Rust";

    fn opus_magnum(guru: &Magister) {
        
        guru.dominari_materia("Fiat Lux");
    }
}

in_umbra_igitur_pugnabimus!(Architectus);
*/